### PR TITLE
[Exactly-Once State Cleanup] Commit

### PR DESCRIPTION
**Pull request by Paultagoras**
## Summary
This should provide more resiliency when we encounter state errors - rather than just throwing an exception, we warn + ignore (mostly). This also changes how we handle when the offset = 0, because is most cases that makes the topic was erased + recreated.

## Checklist
- [ ] Unit and integration tests covering the common scenarios were added
- [ ] A human-readable description of the changes was provided to include in CHANGELOG
- [ ] For significant changes, documentation in https://github.com/ClickHouse/clickhouse-docs was updated with further explanations or tutorials

### CODE REVIEW

File: src/main/java/com/clickhouse/kafka/connect/sink/processing/Processing.java [hunk 1]
```diff
@@ -104,25 +104,28 @@ public void doLogic(List<Record> records) {
...
                 int bpBeforeDrop = records.size();
-                records = dropRecords(stateRecord.getMinOffset(), records);
-                int bpAfterDrop = records.size();
+                trimmedRecords = dropRecords(stateRecord.getMinOffset(), records);
+                int bpAfterDrop = trimmedRecords.size();
                 LOGGER.debug(String.format("before drop %d after drop %d state %s",  bpBeforeDrop, bpAfterDrop, stateRecord.getOverLappingState(rangeContainer)));
                 // Here there are several options
                 switch (stateRecord.getOverLappingState(rangeContainer)) {
+                    case ZERO: // Reset if we're at a 0 state
+                        LOGGER.warn(String.format("It seems you deleted the topic - resetting state for topic [%s] partition [%s].", topic, partition));
```
Comment (mshustov) [inline src/main/java/com/clickhouse/kafka/connect/sink/processing/Processing.java:113]:
```suggestion
                        LOGGER.warn(String.format("The topic seems to be deleted. Resetting state for topic [%s] partition [%s].", topic, partition));
```

File: src/main/java/com/clickhouse/kafka/connect/sink/kafka/RangeContainer.java [hunk 2]
```diff
@@ -42,18 +42,29 @@ public boolean isInRange(long offset) {
...
+        long actualMinOffset = rangeContainer.getMinOffset();
+        long actualMaxOffset = rangeContainer.getMaxOffset();
+
         // SAME State [0,10] Actual [0,10]
-        if (maxOffset == rangeContainer.getMaxOffset() && minOffset == rangeContainer.getMinOffset())
+        if (maxOffset == actualMaxOffset && minOffset == actualMinOffset)
             return RangeState.SAME;
         // NEW State [0,10] Actual [11,20]
-        if (maxOffset < rangeContainer.minOffset)
+        if (actualMinOffset > maxOffset)
```
Comment (mshustov) [inline src/main/java/com/clickhouse/kafka/connect/sink/kafka/RangeContainer.java:58]: We should probably add a diagram with the different states that you created earlier

File: src/main/java/com/clickhouse/kafka/connect/sink/processing/Processing.java [hunk 3]
```diff
@@ -104,25 +104,28 @@ public void doLogic(List<Record> records) {
...
                     case NEW:
-                        doInsert(records);
+                        doInsert(trimmedRecords);
                         stateProvider.setStateRecord(new StateRecord(topic, partition, rangeContainer.getMaxOffset(), rangeContainer.getMinOffset(), State.AFTER_PROCESSING));
                         break;
                     case CONTAINS: // The state contains the given records
-                        // Do nothing - write to dead letter queue
-                        records.forEach( r ->
-                                sendTODlq(record, new DuplicateException(String.format(record.getTopicAndPartition())))
-                        );
```
Comment (mzitnik) [inline src/main/java/com/clickhouse/kafka/connect/sink/processing/Processing.java:121]: Why did you remove sending to DLQ

File: src/test/java/com/clickhouse/kafka/connect/sink/processing/ProcessingTest.java [hunk 4]
```diff
@@ -170,23 +170,23 @@ public void ProcessSplitNewWithBeforeProcessingTest() {
...
+//        Processing processing = new Processing(stateProvider, dbWriter, er);
+//        processing.doLogic(records);
+//        assertEquals(records.size(), dbWriter.recordsInserted());
+//        StateRecord stateRecord = stateProvider.getStateRecord("test", 1);
+//        stateRecord.setState(State.BEFORE_PROCESSING);
+//        processing.doLogic(containsRecords);
+//        assertEquals(records.size(), dbWriter.recordsInserted());
+//        // Check that we send the correct amount of records to DLQ
+//        assertEquals(containsRecords.size(), er.size());
+//    }
```
Comment (mzitnik) [inline src/test/java/com/clickhouse/kafka/connect/sink/processing/ProcessingTest.java:190]: What happened here?

**Comment by mshustov** [comment]
@Paultagoras could you talk to @Ryado to align on the desired behavior?

File: src/main/java/com/clickhouse/kafka/connect/sink/kafka/RangeContainer.java (hunk 2 above)
Comment (mzitnik) [reply src/main/java/com/clickhouse/kafka/connect/sink/kafka/RangeContainer.java:58]: https://github.com/ClickHouse/clickhouse-kafka-connect/blob/main/docs/DESIGN.md 
we just need to add to current system

**Review by mzitnik — APPROVED** [review]
